File: crates/core/tests/fixtures/cs/OrderWorkflow.cs
namespace App.Order;

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

// cm: def OrderStatus
public enum OrderStatus
{
	Draft,
	Paid,
	Cancelled
}

// cm: def Order
public record Order(string Id, OrderStatus Status, DateTimeOffset CreatedAt);

// cm: def IOrderRepository
public interface IOrderRepository
{
	Task<IReadOnlyList<Order>> FindExpiredAsync(DateTimeOffset now, CancellationToken token);
	Task CancelAsync(string id, CancellationToken token);
}

// cm: def OrderWorkflow
public sealed class OrderWorkflow
{
	private readonly IOrderRepository _repository;
	private readonly TimeProvider _timeProvider;

	public OrderWorkflow(IOrderRepository repository, TimeProvider timeProvider)
	{
		_repository = repository;
		_timeProvider = timeProvider;
	}

	// cm: def OrderWorkflow.CancelExpiredAsync
	public async Task<int> CancelExpiredAsync(CancellationToken token)
	{
		// cm: ref OrderWorkflow.CancelExpiredAsync.calls.FindExpiredAsync
		var expired = await _repository.FindExpiredAsync(_timeProvider.GetUtcNow(), token);
		foreach (var order in expired.Where(o => o.Status == OrderStatus.Draft))
		{
			await _repository.CancelAsync(order.Id, token);
		}
		return expired.Count;
	}
}

File: crates/core/tests/fixtures/extractors/java/OrderWorkflow.java
package app.order;

import java.time.Clock;
import java.time.Instant;
import java.util.List;

// cm: def OrderWorkflow
public final class OrderWorkflow implements Runnable {
	private final OrderRepository repository;
	private final Clock clock;

	public OrderWorkflow(OrderRepository repository, Clock clock) {
		this.repository = repository;
		this.clock = clock;
	}

	@Override
	// cm: def OrderWorkflow.run
	public void run() {
		// cm: ref OrderWorkflow.run.calls.findExpired
		List<Order> expired = repository.findExpired(clock.instant());
		expired.forEach(order -> repository.cancel(order.id()));
	}

	// cm: def OrderWorkflow.Status
	public enum Status {
		DRAFT,
		PAID,
		CANCELLED
	}

	// cm: def OrderWorkflow.Order
	public record Order(String id, Status status, Instant createdAt) {}

	// cm: def OrderWorkflow.OrderRepository
	public interface OrderRepository {
		List<Order> findExpired(Instant now);
		void cancel(String id);
	}
}
